File: descripe/imapp_modisl2/src/leocat/src/MOD_PRAlg29/spatial_var.c
/*
***********************************************************************
!F77

!DESCRIPTION:
   Routine for finding the number of pixels surrounding the center
   pixel in a context (nlcntx*necntx pixels) that satisfy a difference
   criterion.

!Input Parameters:
   None

   Inputs stored in structure variable 'pxin' of type "pixel_in" defined
   in pixel.h.

!Output Parameters:
   None

   Outputs stored in structure variable 'rg_var' of type "regional_var"
   defined in pixel.h

!Revision History:
   06/04 Collection 5  R. Frey
   Updated argument list.
   11/07 Converted to C   R. Frey

!Team-unique Header:

!References and Credits:
 See Cloud Mask ATBD  
 
!END
***********************************************************************/


/* Includes */

#include <stdio.h>
#include <math.h>
#include "pixel.h"
#include "thresholds.h"
#include "mask_processing_constants.h"


int spatial_var()


{

/*  Declarations */

	int i;
    int ipt;
    int result;

/*  Initializations */

//  printf("Executing spatial_var \n");
    
    ipt = 0;

/**********************************************************************/


//  Compare surrounding differences to threshold.
    for ( i=0; i<num_diffs_surr; ++i) {

      if ( rintf(rg_var.diff[i]) != rintf(bad_data) ) {
        if ( rg_var.diff[i] <= dovar11[0] ) {
          ipt++;
        }
      }

    }
    rg_var.num_small_diffs = ipt;

//  If all surrounding pixel differences were less than the threshold
//  value, the scene is declared to be uniform.
    if (ipt == num_diffs_surr) {
      result = 1;
    }
    else {
      result = 0;
    }

//  printf("ipt, result %d %d \n", ipt, result);
    
    return result;

}

File: descripe/imapp_modisl2/src/cloudmask/src/spatial_var.f
      subroutine spatial_var(diff,ipt,result)

      implicit none
      save

c---------------------------------------------------------------------
C!F77
c
c!DESCRIPTION:
c     routine for performing spatial variability tests and
c     setting the appropriate bit flag.
c
c!Input Parameters:
c diff          Array of surrounding pixel reflectance or brightness
c               temperature differences from center pixel value
c
c!Output Parameters:
c ipt           number pixels satisfying variability test (8 max)
c results	results of spatial variability test (1=uniform)
c
c!Revision History:
c 06/04 Collection 5  R. Frey
c Updated argument list.
c
c!Team-unique Header:
c
c!References and Credits:
c See Cloud Mask ATBD-MOD-06.  
c 
c!END
c---------------------------------------------------------------------

      include 'global.inc'

      integer masir11
      parameter (masir11=2)
c ...
c ... scalar arguments ..
      integer ipt,result
c ...
c ... array arguments ..
      real diff(var_band,8)
c ...
c ... local scalars ..
      integer i
      INCLUDE 'spatial_var_thr.inc'
c ...
c ... Have chose to only test the 11 micron uniformity for now
      ipt = 0
c ... compare surrounding bt differences to threshold
      do 200 i = 1,8
        if (abs(diff(masir11,i)) - bad_data .gt. 0.1)  then
          if (diff(masir11,i).le.dovar11(1)) then
            ipt = ipt + 1
          end if
        endif
  200 continue

c ... if all surrounding pixel differences were less than the
c ... threshold value, scene is declared to be uniform.
      if (ipt.eq.8) then
        result = 1
      else
        result = 0
      end if

      return
      end
